// mongoose-express/dist/src/classes/errorWithCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorDocumentNotFound = exports.ErrorValidationFailed = exports.ErrorInternalServerError = exports.ErrorUnauthorized = exports.ErrorBadInput = void 0;
class ErrorWithCode {
    constructor(code, message, defails = {}) {
        this.code = code;
        this.message = message;
        this.details = defails;
    }
}
exports.default = ErrorWithCode;
class ErrorBadInput extends ErrorWithCode {
    constructor(details = {}) {
        super(400, 'bad user input', details);
    }
}
exports.ErrorBadInput = ErrorBadInput;
class ErrorUnauthorized extends ErrorWithCode {
    constructor(details = {}) {
        super(403, 'unauthorized', details);
    }
}
exports.ErrorUnauthorized = ErrorUnauthorized;
class ErrorInternalServerError extends ErrorWithCode {
    constructor(details = {}) {
        super(500, 'internal server error', details);
    }
}
exports.ErrorInternalServerError = ErrorInternalServerError;
class ErrorValidationFailed extends ErrorWithCode {
    constructor(details = {}) {
        super(400, 'validation of input data failed', details);
    }
}
exports.ErrorValidationFailed = ErrorValidationFailed;
class ErrorDocumentNotFound extends ErrorWithCode {
    constructor(details = {}) {
        super(404, 'document not found', details);
    }
}
exports.ErrorDocumentNotFound = ErrorDocumentNotFound;
