// mongoose-express/dist/src/classes/wrapperCreateOptions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const wrapperOptions_1 = __importDefault(require("./wrapperOptions"));
class WrapperCreateOptions extends wrapperOptions_1.default {
    constructor(input) {
        var _a, _b, _c, _d;
        super(input);
        this.options = {
            returnCreatedObject: ((_a = input.options) === null || _a === void 0 ? void 0 : _a.returnCreatedObject) === true,
            populateFields: ((_b = input.options) === null || _b === void 0 ? void 0 : _b.populateFields) || [],
            middleware: ((_c = input.options) === null || _c === void 0 ? void 0 : _c.middleware)
                ? {
                    create: input.options.middleware.create || []
                } : {
                create: []
            },
            postProcessing: ((_d = input.options) === null || _d === void 0 ? void 0 : _d.postProcessing) || []
        };
        this.mongooseModel = input.mongooseModel
            ? Object.assign({}, input.mongooseModel, {
                MongoBouncer: input.mongooseModel.MongoBouncer ? input.mongooseModel.MongoBouncer : { Disabled: true }
            })
            : { MongoBouncer: { Disabled: true } };
    }
}
exports.default = WrapperCreateOptions;
