// mongoose-express/dist/src/classes/wrapperFindOptions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const wrapperOptions_1 = __importDefault(require("./wrapperOptions"));
class WrapperFindOptions extends wrapperOptions_1.default {
    constructor(input) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        super(input);
        this.options = {
            initialQuery: ((_a = input.options) === null || _a === void 0 ? void 0 : _a.initialQuery) || {},
            exclude: ((_b = input.options) === null || _b === void 0 ? void 0 : _b.exclude) || [],
            populateFields: ((_c = input.options) === null || _c === void 0 ? void 0 : _c.populateFields) || [],
            middleware: ((_d = input.options) === null || _d === void 0 ? void 0 : _d.middleware)
                ? {
                    find: input.options.middleware.find || []
                } : {
                find: []
            },
            paramsForQuery: ((_e = input.options) === null || _e === void 0 ? void 0 : _e.paramsForQuery) || [],
            postProcessing: ((_f = input.options) === null || _f === void 0 ? void 0 : _f.postProcessing) || []
        };
        if ((_g = input.options) === null || _g === void 0 ? void 0 : _g.limit)
            this.options.limit = input.options.limit;
        if ((_h = input.options) === null || _h === void 0 ? void 0 : _h.skip)
            this.options.skip = input.options.skip;
        if ((_j = input.options) === null || _j === void 0 ? void 0 : _j.sort)
            this.options.sort = input.options.sort;
        this.mongooseModel = input.mongooseModel
            ? Object.assign({}, input.mongooseModel, {
                MongoBouncer: input.mongooseModel.MongoBouncer ? input.mongooseModel.MongoBouncer : { Disabled: true }
            })
            : { MongoBouncer: { Disabled: true } };
    }
}
exports.default = WrapperFindOptions;
