// mongoose-express/dist/src/classes/wrapperFindOneByIdAndUpdateOptions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const wrapperOptions_1 = __importDefault(require("./wrapperOptions"));
class WrapperUpdateOptions extends wrapperOptions_1.default {
    constructor(input) {
        var _a, _b, _c, _d, _e, _f;
        super(input);
        this.options = {
            initialItem: ((_a = input.options) === null || _a === void 0 ? void 0 : _a.initialItem) || {},
            exclude: ((_b = input.options) === null || _b === void 0 ? void 0 : _b.exclude) || [],
            populateFields: ((_c = input.options) === null || _c === void 0 ? void 0 : _c.populateFields) || [],
            middleware: ((_d = input.options) === null || _d === void 0 ? void 0 : _d.middleware)
                ? {
                    findByIdAndUpdate: input.options.middleware.findByIdAndUpdate || []
                } : {
                findByIdAndUpdate: []
            },
            postProcessing: ((_e = input.options) === null || _e === void 0 ? void 0 : _e.postProcessing) || []
        };
        if ((_f = input.options) === null || _f === void 0 ? void 0 : _f.updateFields)
            this.options.updateFields;
        this.mongooseModel = input.mongooseModel
            ? Object.assign({}, input.mongooseModel, {
                MongoBouncer: input.mongooseModel.MongoBouncer ? input.mongooseModel.MongoBouncer : { Disabled: true }
            })
            : { MongoBouncer: { Disabled: true } };
    }
}
exports.default = WrapperUpdateOptions;
