// mongoose-express/dist/test/helpers/routeTester.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
require("express");
const sinon_1 = __importDefault(require("sinon"));
class RouteTester {
    constructor(router, req) {
        let resolveHasEnded;
        this.req = req || {};
        this.req.params = (req === null || req === void 0 ? void 0 : req.params) || {};
        this.req.query = (req === null || req === void 0 ? void 0 : req.query) || {};
        this.req.body = (req === null || req === void 0 ? void 0 : req.body) || {};
        this.req.headers = (req === null || req === void 0 ? void 0 : req.headers) || {};
        this.hasEnded = new Promise((res) => {
            resolveHasEnded = res;
        });
        this.res = {
            status: sinon_1.default.stub(),
            json: sinon_1.default.stub(),
            send: sinon_1.default.stub(),
            end: sinon_1.default.stub()
        };
        this.res.status.returns(this.res);
        this.res.json.returns(this.res);
        this.res.send.returns(this.res);
        this.res.end.returns(this.res);
        this.res.json.callsFake(() => resolveHasEnded(true));
        this.res.send.callsFake(() => resolveHasEnded(true));
        this.res.end.callsFake(() => resolveHasEnded(true));
        this.router = router;
    }
    async execute(opts) {
        this.router.handle(Object.assign({}, this.req, { method: opts.method, url: opts.url }), this.res);
        await this.hasEnded;
    }
}
exports.default = RouteTester;
