// mongoose-express/dist/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ControllerGetByQuery = exports.ControllerGetById = exports.ControllerGetAll = exports.ControllerFindByIdAndUpdate = exports.ControllerDeleteById = exports.ControllerCreate = void 0;
const controller__default_create_1 = __importDefault(require("./controllers/main/controller._default.create"));
exports.ControllerCreate = controller__default_create_1.default;
const controller__default_deleteById_1 = __importDefault(require("./controllers/main/controller._default.deleteById"));
exports.ControllerDeleteById = controller__default_deleteById_1.default;
const controller__default_findByIdAndUpdate_1 = __importDefault(require("./controllers/main/controller._default.findByIdAndUpdate"));
exports.ControllerFindByIdAndUpdate = controller__default_findByIdAndUpdate_1.default;
const controller__default_getAll_1 = __importDefault(require("./controllers/main/controller._default.getAll"));
exports.ControllerGetAll = controller__default_getAll_1.default;
const controller__default_getById_1 = __importDefault(require("./controllers/main/controller._default.getById"));
exports.ControllerGetById = controller__default_getById_1.default;
const controller__default_getByQuery_1 = __importDefault(require("./controllers/main/controller._default.getByQuery"));
exports.ControllerGetByQuery = controller__default_getByQuery_1.default;
